retention/policy_resolver.go

```diff
@@ -0,0 +1,222 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package retention
+
+import (
+	"errors"
+	"time"
+
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/time"
+
+	"github.com/facebookgo/clock"
+)
+
+var (
+	errInvalidRange = errors.New("invalid range; from does not precede until")
+)
+
+// A PolicyTimeRange is the application of a retention Policy within a given
+// time range
+type PolicyTimeRange struct {
+	xtime.Range
+	Policy
+}
+
+// PolicyResolverOptions are options for creating a PolicyResolver
+type PolicyResolverOptions interface {
+	// Logger is the resolver's logger
+	Logger(logger xlog.Logger) PolicyResolverOptions
+	GetLogger() xlog.Logger
+
+	// Clock is the resolver's clock
+	Clock(c clock.Clock) PolicyResolverOptions
+	GetClock() clock.Clock
+}
+
+// NewPolicyResolverOptions returns a new set of PolicyResolverOptions
+func NewPolicyResolverOptions() PolicyResolverOptions {
+	return new(policyResolverOptions)
+}
+
+// A PolicyResolver resolves policies based on a set of retention rules
+// TODO(mmihic): This should be driven off a RuleProvider, shard, etc.
+type PolicyResolver interface {
+	// ResolveTimeline resolves the retention Policy timeline for a given id
+	ResolveTimeline(id string, from, until time.Time) ([]PolicyTimeRange, error)
+}
+
+// NewPolicyResolver returns a new PolicyResolver given a rule provider
+func NewPolicyResolver(p RuleProvider, opts PolicyResolverOptions) (PolicyResolver, error) {
+	if opts == nil {
+		opts = NewPolicyResolverOptions()
+	}
+
+	c, log := opts.GetClock(), opts.GetLogger()
+	if c == nil {
+		c = clock.New()
+	}
+
+	if log == nil {
+		log = xlog.NullLogger
+	}
```

Comment:
The other pattern we're using is to add a `Validate() error` method that we call on options inside of the constructor, i.e.:

```go
func NewThing(opts Options) (Thing, error) {
  if opts == nil {
    opts = NewOptions()
  }
  if err := opts.Validate(); err != nil {
    return nil, err
  }
  return &thing{/*... use opts freely ...*/}
}